Rosetta: Retrieve account sequence number

Maelkum opened the pull request:
This PR introduces functionality that was previously missing - retrieving sequence number for an account key, required for the Rosetta Construction API

## Checklist

- [x] Is on the right branch
- [ ] ~Documentation is up-to-date~
- [ ] ~Tests are up-to-date~

awfm9 commented on api/rosetta/retriever.go:
```diff
@@ -27,4 +27,5 @@ type Retriever interface {
 	Block(rosBlockID identifier.Block) (*object.Block, []identifier.Transaction, error)
 	Transaction(rosBlockID identifier.Block, rosTxID identifier.Transaction) (*object.Transaction, error)
 	Balances(rosBlockID identifier.Block, rosAccountID identifier.Account, rosCurrencies []identifier.Currency) (identifier.Block, []object.Amount, error)
+	AccountSequenceNumber(rosAccountID identifier.Account, keyIndex int) (uint64, error)
```
`SequenceNumber` is enough. We already provide an account as input, so of course it's for an account. Actually, it's for an account and a key index. So why not call it `AccountKeySequenceNumber`? Because it's not necessary :D

awfm9 commented on rosetta/retriever/invoker.go:
```diff
@@ -16,8 +16,10 @@ package retriever
 
 type Invoker interface {
 	Script(height uint64, script []byte, parameters []cadence.Value) (cadence.Value, error)
+	GetAccount(address flow.Address, height uint64) (*flow.Account, error)
```
Nooooo. Please re-read https://golang.org/doc/effective_go#Getters.

What's the advantage of `invoke.GetAccount` over `invoke.Account`?

Maelkum replied:
I was just following suite with the existing code that was `GetAccount` 😃 
I've renamed `GetAccount` to `Account` in other places too.

awfm9 commented on rosetta/retriever/retriever.go:
```diff
@@ -387,3 +388,50 @@ func (r *Retriever) operations(txID flow.Identifier, events []flow.Event) ([]*ob
+	}
+
+	// Lookup the specified key.
+	for _, key := range account.Keys {
```
Are you sure this is needed and they are not already sorted by index? If that's the case, I would much prefer creating a lookup map for the keys by index first, and then checking just the index we want, rather than iterating through all of them. Yes, it's a bit less efficient, but that's not where we will save relevant performance. The code will be nicer, though.

Maelkum replied:
I'm not sure that they aren't already sorted. Still, the fact that there's a field titled `Index` in the structure made me feel uneasy about making any assumptions, and I thought it's best to be explicit in checking.

I've changed the code to use a lookup map now, I think the function reads more direct now.

awfm9 reviewed (commented):
I changed my mind on the cache for the sequence number. Due to the way the Construction API is defined by Rosetta, it makes it impossible to proper track sequence numbers across the various endpoints. As such, I think we should leave it to the consumer of the API to track transaction status.

awfm9 reviewed (changes requested):
I just noticed that you use the Flow Go SDK model for an address; we should absolutely not use that. Always the core Flow models. If that's the case anywhere else, please fix it too!

Maelkum commented:
> I just noticed that you use the Flow Go SDK model for an address; we should absolutely not use that. Always the core Flow models. If that's the case anywhere else, please fix it too!

> I'm sorry I didn't see this earlier, but we should avoid any dependency on the Flow Go SDK, and use the native Flow models instead: github.com/onflow/flow-go/model/flow.

@awfm9 
I normally do use Flow native types.

However, at least to some extent, I think we need the SDK types. For example, I use the SDK `Client` to submit transaction for execution (using the [SendTransaction](https://pkg.go.dev/github.com/onflow/flow-go-sdk/client#Client.SendTransaction)).
Then the transaction type is also the SDK [Transaction](https://pkg.go.dev/github.com/onflow/flow-go-sdk#Transaction), not the Flow [TransactionBody](https://pkg.go.dev/github.com/onflow/flow-go@v0.20.6/model/flow#TransactionBody).

That's why the code that explicitly deals with transactions (e.g. `CompileTransaction`, `ParseTransaction`) works with the SDK types, which is what is in the transaction itself.

I can change it for the errors, but we'd still be doing coversion between SDK and Flow types.

Alternatively, I could always work with the Flow types, until the very end, in the `Submit` endpoint, at which point I convert from `flow.TransactionBody` to the `flow-go-sdk` `Transaction` type, and submit that.

awfm9 commented:
> However, at least to some extent, I think we need the SDK types. For example, I use the SDK `Client` to submit transaction for execution (using the [SendTransaction](https://pkg.go.dev/github.com/onflow/flow-go-sdk/client#Client.SendTransaction)).
> Then the transaction type is also the SDK [Transaction](https://pkg.go.dev/github.com/onflow/flow-go-sdk#Transaction), not the Flow [TransactionBody](https://pkg.go.dev/github.com/onflow/flow-go@v0.20.6/model/flow#TransactionBody).
> 
> That's why the code that explicitly deals with transactions (e.g. `CompileTransaction`, `ParseTransaction`) works with the SDK types, which is what is in the transaction itself.

Ooof. That's unfortunate. In that case, we should at least use the native Flow address type, i.e. in errors, where it doesn't need to be used for the SDK.

Maelkum commented:
@awfm9 
I have removed the SDK types from where they're not needed. The *only* places where we still reference the SDK types are those that explicitly touch the SDK `Transaction`:
- `api/rosetta` - endpoints that deal with creating, parsing, submitting or returning the ID of the transaction (`hash` endpoint)
- `rosetta/transaction` - deals with compiling and parsing of transactions. I've moved the conversion between the SDK `Address` and native Flow `Address` types to a `convert.go` file, so we have only one file where we import both packages
- `rosetta/submitter` - uses the SDK client to submit the transaction
- `flow-rosetta-server/main` - creation of the SDK client for the submitter to use.

awfm9 commented on rosetta/failure/invalid_proposal_key.go:
```diff
@@ -0,0 +1,31 @@
+}
+
+func (i InvalidProposalKey) Error() string {
+	return fmt.Sprintf("invalid proposal key (address: %s, key index: %d): %s", i.Address.String(), i.Index, i.Description)
```
You know that `%s` automatically uses the `String()` function, right? :D

Maelkum replied:
That used to be a `%v`, but I changed it to be explicit, forgot that one :D
